File: RB101_RB109_small_problems/medium_2/now_i_know_my_abcs.rb
=begin

PEDAC Template
==============

(Understand the) Problem
------------------------

-  Inputs: string
-  Output: boolean representing if the input string can be spelled with the spelling blocks provided

---

**Problem Domain:**

Spelling blocks are broking into the following groups:

B:O   X:K   D:Q   C:P   N:A
G:T   R:E   F:S   J:W   H:U
V:I   L:Y   Z:M

Each letter in each block can only be used once.

---

**Explicit Requirements:**
- Each letter in each block can only be used once.
---


**Implicit Requirements:**
- Method should be named #block_word?
- Letter case does not matter (case insensitive)
---

**Clarifying Questions:**

1. Can letters be repeated? Example: would 'apple' return true?
  No. 

---

**Mental Model:**

Iterate through the letters in the input string. 
If the iteration's letter has not been used yet (i.e. the block_tracker hash's value for the associate key is false), flip the block_tracker hash's associated value to true. 
Otherwise, return false.
If all iterations complete without returning false, return true.
---

Examples / Test Cases / Edge Cases
----------------------------------

**Examples:**

block_word?('BATCH') == true
block_word?('BUTCH') == false
block_word?('jest') == true

---

Data Structure
--------------

Track which blocks are used with a hash. The keys are strings of each character on the block. The value will be a boolean representing if the block has been used. Initially all values will be false, but a value will be toggled to true once a character on the associated block is used.

block_tracker = {
  'bo' => false,
  'xk' => false,
  'dq' => false,
  etc..
}

Create block_tracker_keys array to convert from letter in string to the appropriate hash key.

block_tracker_keys = ['bo', 'xk', 'dq', etc...]

---

Algorithm
---------

- Create block tracker hash with a default value of false. Assign to variable 'block_tracker'.
- Create array with the keys from 'block_tracker'. Assign to variable 'block_tracker_keys'
- Iterate through the input string's letters case insensitively:
  - Determine the appropriate 'block_tracker' hash key from the 'block_tracker_keys' array. Assign to variable 'key'.
  - If 'block_tracker[key]' is false, set it to true. Otherwise, return false.
- Return true.

Code
----

=end

def block_word?(str)
  block_tracker_keys = %w(bo xk dq cp na gt re fs jw hu vi ly zm)
  
  block_tracker = Hash.new(false)
  
  str.downcase.split('').each do |letter|
    key = nil
    block_tracker_keys.each do |block|
      if block.include?(letter)
        key = block
      end
    end
    
    if block_tracker[key]
      return false
    else
      block_tracker[key] = true
    end
  end
  true
end

p block_word?('BATCH') == true
p block_word?('BUTCH') == false
p block_word?('jest') == true

#LS Solution
BLOCKS = %w(BO XK DQ CP NA GT RE FS JW HU VI LY ZM).freeze

def block_word?(string)
  up_string = string.upcase
  BLOCKS.none? { |block| up_string.count(block) >= 2 }
end

File: JS210_small_problems/interpretive_problem_solving/now_i_know_my_abcs.js
/*

A collection of spelling blocks has two letters per block, as shown in this list:

B:O   X:K   D:Q   C:P   N:A
G:T   R:E   F:S   J:W   H:U
V:I   L:Y   Z:M

This limits the words you can spell with the blocks to only those words that do not use both letters from any given block. You can also only use each block once.

Write a function that takes a word string as an argument and returns true if the word can be spelled using the set of blocks, false otherwise. You can consider the letters to be case-insensitive when you apply the rules.

PEDAC Template
==============

(Understand the) Problem
------------------------

-  Inputs: string representing a word
-  Output: return a boolean representing whether the word input can be spelled with the spelling blocks

---

**Problem Domain:**
Nothing specific. See problem statement above.
---

**Explicit Requirements:**
- Function must check if the word string argument can be spelled using a single spelling block no more than once.
- Each block can only be used once. 
- Spelling blocks have two letters
- Consider letters to be case-insensitive when applying rules.
- Must return a boolean representing whether the word input can be spelled with the spelling blocks
---


**Implicit Requirements:**
- Function must be named isBlockWord
---

**Clarifying Questions:**

1. Can the same word from a single block be used once? i.e should isBlockWord("DUD") return true?
  No. 
2. Should we assume the input is always a string containing at least one letter?
  Yes.
3. Should we assume the input will always contain only letters?
  No. Do not consider non-letter characters when applying the rules.
---

**Mental Model:**

- Create an array with strings containing each spelling block's letters.
- Clean the input string of any non-letter characters and convert all letters to uppercase.
- Iterate over the uppercased letters in the cleaned input string.
  - If a block has been used already, return false.
  - Mark the appropriate block as used for the current character.
- Return true if iteration completes without returning early.

---

Examples / Test Cases / Edge Cases
----------------------------------

**Examples:**

isBlockWord('BATCH');      // true
isBlockWord('BUTCH');      // false
isBlockWord('jest');       // true
isBlockWord('DUD');       // false
isBlockWord('dUd');       // false
isBlockWord('3BaTCH');    // true
isBlockWord('0=OXEN');    // true

---

Data Structure
--------------

To represent the spelling blocks, use an array like:
[
  'BO',
  'XK',
  'DQ',
  'CP',
  'NA',
  'GT',
  'RE',
  'FS',
  'JW',
  'HU',
  'VI',
  'LY',
  'ZM',
]

To represent whether a spelling block has been used, use an object to track which spelling blocks have been used for a given word. The keys for the object will be the strings in the above array. A boolean will represent whether the block has been used or not.
Ex:
{
  'BO': false,
  'XK': true,
  etc...
}

---

Algorithm
---------

- Create an array with strings containing each spelling block's letters.
- Clean the input string of any non-letter characters and convert all letters to uppercase.
  - Use a regex like /[^a-z]/ig to replace non-letters with empty strings.
  - Use String.prototype.toUpperCase() to convert all letters to uppercase.
- Iterate over the uppercased letters in the cleaned input string.
  - If a block has been used already, return false.
  - Mark the appropriate block as used for the current character.
- Return true if iteration completes without returning early.

Code
----
*/

function cleanInputString(str) {
  const NON_LETTER_REGEX = /[^a-z]/ig;

  return str.replaceAll(NON_LETTER_REGEX, '').toUpperCase();
}

function isBlockWord(inputStr) {
  const SPELLING_BLOCKS = [
                            'BO',
                            'XK',
                            'DQ',
                            'CP',
                            'NA',
                            'GT',
                            'RE',
                            'FS',
                            'JW',
                            'HU',
                            'VI',
                            'LY',
                            'ZM',
                          ];

  let blockUseTracker = {};

  let cleanedInput = cleanInputString(inputStr);

  for (let char of cleanedInput) {
    let block = SPELLING_BLOCKS.filter(blockLetters => blockLetters.includes(char))[0];

    if (blockUseTracker[block]) return false;
  
    blockUseTracker[block] = true;
  }

  return true;
}

console.log(isBlockWord('BATCH'));      // true
console.log(isBlockWord('BUTCH'));      // false
console.log(isBlockWord('jest'));       // true
console.log(isBlockWord('DUD'));       // false
console.log(isBlockWord('dUd'));       // false
console.log(isBlockWord('3BaTCH'));    // true
console.log(isBlockWord('0=OXEN'));    // true
console.log(isBlockWord('APPLE'));      // false
console.log(isBlockWord('apple'));      // false
console.log(isBlockWord('apPLE'));      // false
console.log(isBlockWord('Box'));        // false